text/0045-add-notion-of-protocol.md

```diff
@@ -0,0 +1,143 @@
+# Summary
+- Define supported protocols for services and outputs in libobs and OBS Studio
+- Outputs for streaming are registered with one or more compatible protocols and codecs
+- Only codecs compatible with the output and the service are shown
+- Audio codec can be chosen if various compatible, the encoder implementation can be chosen in advanced output mode.
+- Will be extended with RFC 39
+
+# Motivation
+Create a better management of outputs, encoders, and services with an approach focused on protocols.
+
+# Design
+
+## Outputs API
+Only outputs with the `OBS_OUTPUT_SERVICE` flag are concerned.
+
+Adding to `obs_output_info`:
+- `const char *protocols`: protocols separated with a semi-colon (ex: `"RTMP;RTMPS"`), required to register the output.
+- `const char *protocols_prefixes`: URL protocols prefixes separated with a semi-colon `"rtmp://;rtmps://"`, optional.
```

Comment:
It's always optional, adding it just allow the protocol to be auto-detected.

Reply:
Allow me to rephrase. In what cases would one provide prefixes? In what cases would one _not_ provide prefixes?

Reply:
> In what cases would one provide prefixes?

If the protocol has a unique prefix (e.g. RTMP has `rtmp://`, SRT has`srt://`,…)

> In what cases would one not provide prefixes?

Protocols relying on HTTP will be a common case (e.g. HLS, WHIP).

I made it optional because of those, the prefix needs to be unique to the protocol.

Reply:
For posterity... As explained off-thread, in my opinion, outputs and services should be able to specify all relevant information about them. It is up to the consumer of that information to decide how to act on that information. For example, if we determine that there are multiple outputs that all use the same protocol prefix, then we should not automatically provide whatever convenience functionality we would if the protocol prefix were unique to all registered outputs. We (and other obs_output_info creators) should not have to be careful with specifying obs_output_info to ensure that protocols_prefixes are unique. The consumer of that data should determine if a given protocol prefix is unique and can be safely mapped to a single protocol or obs_output_info.

In that line of thinking, all obs_output_info objects should always fully specify their `protocols` and `protocols_prefixes`, and OBS (whether in obs-outputs or somewhere else) should figure out if it is safe to provide convenience functionality based on uniqueness (i.e., auto-detecting the protocol based on the prefix and adjusting the UI accordingly) or if it must provide a different user experience 

I don't know if this is something we can, or should, address now in this RFC, or if this is something that can easily be addressed later on. cc @PatTheMav who might be able to better explain this point.

Reply:
Following further off-thread discussion, the `protocols_prefixes` items were removed from the RFC and related PRs. Resolving.